### ISSUE TITLE
24fps (noticeable) when opening animation demo on Nexus 5

### ISSUE THREAD

**Issue by sethladd** [flutter/flutter#9219]
There is a noticeable un-smoothness when opening the animation demo.

Make a release build of the gallery, open animation demo:

![screenshot_20170405-103110](https://cloud.githubusercontent.com/assets/5479/24718542/37e706f6-19eb-11e7-89b1-578cc00c55ef.png)

**Comment by sethladd** [comment]
For reference, here are the spikes when opening pesto, shrine, and contacts:

![screenshot_20170405-103355](https://cloud.githubusercontent.com/assets/5479/24718593/72d1d35e-19eb-11e7-92af-8987e0f7b51c.png)

**Comment by eseidelGoogle** [comment]
Seth believes this should block an updated release of the gallery.  The next build is slated to being release verification next week, so we'd need to fix this by then.

**Comment by sethladd** [comment]
On a Moto G4 (the lower-bound of our Android phone target) we see 22fps when opening any component/widget demo (not just the vignettes)

![screenshot_20170406-132553](https://cloud.githubusercontent.com/assets/5479/24774312/4dd1efe8-1acd-11e7-9383-d30e2999ae36.png)

**Comment by eseidelGoogle** [comment]
@chinmaygarde do you know if we upload images async to the GPU?  Just musing at Triage about why these huge spikes might exist.

**Comment by chinmaygarde** [comment]
It is the raster cache snapshotting huge portions of the hierarchy when the animation is about to start and settle at either end. This work is expensive and completely unnecessary.

![screen shot 2017-04-11 at 3 31 35 pm](https://cloud.githubusercontent.com/assets/44085/24933624/123898e4-1ecc-11e7-92ae-3456b7bf5ef6.png)

**Comment by chinmaygarde** [comment]
I investigated this further. Though the extra snapshotting work does take time, it is not the bottleneck. The `SkCanvas::Flush` itself takes a long time. This suggests we are simply making too many draw calls to the GPU. We need to simplify the scene.

![screen shot 2017-04-11 at 4 45 49 pm](https://cloud.githubusercontent.com/assets/44085/24935364/72aff834-1ed6-11e7-966d-4d1109147e91.png)

**Comment by ScottPierce** [comment]
@chinmaygarde So that means that the scene has hit the limit of what Flutter can do on the phone? Are there any other potential areas of optimization?

**Comment by eseidelGoogle** [comment]
@chinmaygarde so next step is to look at the skpicture?  Or where would we look next for reducing draw calls?  What sort of tooling/logging should we add to understand situations like this?

**Comment by Hixie** [comment]
In that screenshot, we're being descheduled for a really long time. The flush is long, but not THAT long.

Can we break down what's in the scene and how expensive each operation is? It doesn't seem that complicated superficially. It has very little text, for instance. A few gradients, a few rectangles, some images.

**Comment by chinmaygarde** [comment]
The last time we saw a trace like this (i.e. one where the flush was
abnormally long but the CPU time was relatively short), we traced the
difference to the GL driver. Also, given that I am seeing wildly different
traces on iOS, Android & Mac Desktop makes me more certain that this is the
issue. Also, if the swap was taking the majority of the frame slice, I
would have suspected it was frame scheduling or pipelining. From the
traces, that is not the case.

Next Steps:

To be absolutely certain, I can add traces to all GL calls and try to add
up the wall times and CPU times from each frame slice (there is a macro in
Skia that allows you to do this, will apply it and post updated traces). I
will also get traces from the Vulkan backend and see how they differ.

I'll also capture an SkPicture and share it with the Skia team to see what
they think.

I am not too confident I will be able to get reliable OpenGL traces on
Android, so the best way to identify how expensive each operation is would
be to keep simplifying the SkPicture incrementally and noting the
difference.

**Comment by sethladd** [comment]
Thanks for doing the research!

**Comment by eseidelGoogle** [comment]
@chinmaygarde is evaluating, if this can't be fixed by EOW we'll disable the demo.

**Comment by chinmaygarde** [comment]
Update: Within the SkCanvas::Flush, it seems to be the frequent render target switches that cause slowdowns. In the attached trace, glClear on a frame with numerous render target switches causes a stall.
[LogGLCalls.json.zip](https://github.com/flutter/flutter/files/931288/LogGLCalls.json.zip)
![screen shot 2017-04-18 at 3 50 03 pm](https://cloud.githubusercontent.com/assets/44085/25156327/bca6d2e6-244e-11e7-8eb8-c81073a4a068.png)

**Comment by chinmaygarde** [comment]
Another data point: The Vulkan backend does not have this issue. Frame rates are much more consistent. This is the timeline when moving in and out of the demo.
![flutter_01](https://cloud.githubusercontent.com/assets/44085/25156699/1159d3ea-2451-11e7-835f-756497f9a791.png)

**Comment by chinmaygarde** [comment]
The render target switches were traced to the excessive use of `OpacityLayer`s. All transitions that happen when we open a demo use a full screen opacity layer. This causes a slowdown on most devices (like the one [described](#issuecomment-292308315) by @sethladd above). Using a simple animation that does not use an opacity transition (like the ones used elsewhere in the Android framework) fixes that issue.

In the case of the "Animation" demo, more opacity layers are used to fade away the images on the tabs. This causes a render target switch per tab. Overlaying a simple color layer on top of the image without the opacity achieves the same visual effect without the corresponding hit in performance.

Bottom line: Use compositor layers that perform a render target switch (via a `SkCanvas::saveLayer`) sparingly.

**Comment by xster** [comment]
We were talking about a dynamic classification method. I think grouping into general year classes is a good way to dynamically toggle settings and adjust graphical fidelity for performance.
go/device-classifier

Let's use the data and hand switch expensive UI features on year class as long as we don't use it too much.

**Comment by eseidelGoogle** [comment]
The resolution I would recommend is that we make a surgical fix to the Animation Demo ( @HansMuller ) given the time constraints and we file separate bugs about enhancements we want to make.

For example, @chinmaygarde do you feel like you had sufficient debugging tools for this?  Should we be surfacing some additional information to catch these cases in the future?

Also, iOS uses transparency all over the place.  Are they doing something fancy that we're not here (probably) or do we need to work with Skia to make this kind of operation less expensive (assuming that's possible)?  Or other options? :D 

Probably all of those above questions/answers should just end up as bugs which relate back to this (giant) one. :)

**Comment by chinmaygarde** [comment]
I have a fix that replaces the [fade transition used on Android](https://github.com/flutter/flutter/blob/af25a5b888052e83ab4b5e4aac0ab111a0bdc135/packages/flutter/lib/src/material/page.dart#L42) with a simple slide out. That fixes perf issues on opening all demos. @xster mentioned that this used to be the previous behavior anyway. So I don't I think simplifying that transition should be an issue.

Specifically for these page transitions, the [CupertinoPageTransition](https://github.com/flutter/flutter/blob/af25a5b888052e83ab4b5e4aac0ab111a0bdc135/packages/flutter/lib/src/material/page.dart#L176) used on iOS does not use opacity layers. So it was always a bit cheaper on iOS. For other uses of opacity layers, we can detect cases where we can get away from using an `SkCanvas::saveLayer` and applying the opacity via a paint. For example, the opacity applied to the items on the tabs can probably be applied in the fragment shader instead of requiring a render target switch (the layer has a single child so group opacity is unnecessary). I will have to investigate how to do that with Skia. But at least now I know what to optimize for.

**Comment by Hixie** [comment]
I would very much prefer if we kept the fade animation... it's what other Android apps do on those devices.

**Comment by chinmaygarde** [comment]
> it's what other Android apps do on those devices.

While @sethladd, @xster and I were tooling around on the device, we saw that the behavior was inconsistent. The contacts app had the slide transition while some parts of Google Play Music had the fade.

In any case, if really like the fade transition, we can also introduce a picture layer boundary at that point in the hierarchy and make that picture be eligible for participation into the raster cache. We would pay the cost of populating the raster cache entry once (and may skip a frame or two on less powerful hardware) but subsequent frames would be fine. I'll submit patches for both tomorrow and we can just keep the one we like.

**Comment by sethladd** [comment]
> I would very much prefer if we kept the fade animation... it's what other Android apps do on those devices.

Once we make it fast, we should add it back. It looks really nice.

**Comment by Hixie** [comment]
I thought Opacity already always introduced a repaint boundary.

**Comment by chinmaygarde** [comment]
> I thought Opacity already always introduced a repaint boundary.

That is not true.

Update:
* [This pull request](https://github.com/flutter/flutter/pull/9472) should make the transitions on all demos fast.
* The perf issue on the Animations demo can be fixed by removing [this opacity widget](https://github.com/flutter/flutter/blob/3cf23fcb202d6cb4377f5d7d1e0d547ec4537c10/examples/flutter_gallery/lib/demo/animation/widgets.dart#L36) and expressing the effect in terms of a custom color and blend mode. `RenderImage` currently hard codes the blend mode to [`BlendMode.srcIn`](https://github.com/flutter/flutter/blob/3cf23fcb202d6cb4377f5d7d1e0d547ec4537c10/packages/flutter/lib/src/rendering/image.dart#L103). @HansMuller is working on making that configurable. He says he will be done with that by today. We can then fix the demo to use the new stuff.
* If we need to release the demo before Hans is able to land the fix, we can manually change the opacity of the images on the section tabs (they are in the asset bundle) and remove the opacity layer. That should give us equivalent performance but is obviously a hack.
* Issues of this class will be caught earlier in the future once we enable a debugging option where offscreen rendered layers are checker-boarded. The issue tracking the implementation of that option has been [filed here](https://github.com/flutter/flutter/issues/9473).
* The particular issue of the tabs in the animations demo (single child in an opacity layer) can also be detected by `flow` and an optimization applied where the render target switch is avoided in favor of achieving the same effect via paint options (like the one we are manually applying in this case). The issue tracking the implementation of that optimization is [filed here](https://github.com/flutter/flutter/issues/9480).

**Comment by Hixie** [comment]
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/rendering/proxy_box.dart#L709
...says that Opacity will generate a new layer whenever the opacity is not 100% or 0%. At 100% it just passes everything through unmodified. At 0% it doesn't render anything.

Is this not working?

**Comment by chinmaygarde** [comment]
Ah, I see. I was looking at [`isRepaintBoundary` on `RenderObject`](https://docs.flutter.io/flutter/rendering/RenderObject/isRepaintBoundary.html). But, even if a picture layer is introduced, the raster cache may disregard it for caching purposes if the op count of the picture is low enough (which it is in this case because there is only child). To force the raster cache to always cache the picture, you would need to set the [`isComplexHint`](https://docs.flutter.io/flutter/rendering/PictureLayer/isComplexHint.html) on the same. I am hesitant to set that hint for this particular layer because doing it correctly via a paint in a `flow` optimization is better than making the raster cache populate an entry for us. Populating the raster cache entry will incur a hit initially and will also take up extra memory.

Ideally, the user will the performance debugging tools to avoid creating unnecessary opacity layers. In case that is not the case, the `flow` optimizations should catch most remaining cases.

**Comment by xster** [comment]
In the case of the transitions, there's only one Opacity widget in the app if I understand correctly right? We should let users have one opacity without having to do performance debugging and removing the opacity.

**Comment by Hixie** [comment]
I don't think we can count on our users doing anything involving profiling FadeTransitions, especially those involved in a MaterialPageRoute. They're going to want to assume that Just Works.

**Comment by chinmaygarde** [comment]
Hmm, even with the `flow` optimizations, we will only be able to handle cases where the fade is applied to a pictures that have a child (or maybe multiple children that themselves have no children, and, don't overlap). Using `FadeTransitions` on nested hierarchies excessively will cause performance bottlenecks. Some profiling there is unavoidable.
